File: py/cd_inversion/training_data.py
"""Generate synthetic CD spectroscopy dataset."""
import numpy as np
from cd_forward_model import WAVELENGTHS, predict_spectrum


def generate_cd_dataset(N: int, noise_sigma: float = 0.5, seed: int = 42):
    rng = np.random.default_rng(seed)
    spectra = np.zeros((N, len(WAVELENGTHS)))
    compositions = np.zeros((N, 3))
    for i in range(N):
        x1, x2 = np.sort(rng.random(2))
        h, s, c = x1, x2 - x1, 1 - x2
        compositions[i] = [h, s, c]
        clean = predict_spectrum(h, s, c)
        spectra[i] = clean + noise_sigma * rng.standard_normal(len(WAVELENGTHS))
    return WAVELENGTHS, spectra, compositions

File: jl/cd_inversion/src/TrainingData.jl
module TrainingData

using ..CDForwardModel
using Random

export generate_cd_dataset

function generate_cd_dataset(N::Int; noise_σ::Float64=0.5, seed::Int=42)
    rng = MersenneTwister(seed)
    params = CDParams()
    wavelengths = params.wavelengths

    spectra      = zeros(N, length(wavelengths))
    compositions = zeros(N, 3)   # [helix, sheet, coil]

    for i in 1:N
        # Sample random composition on the 2-simplex
        h, s, c = _sample_simplex(rng)
        compositions[i, :] = [h, s, c]
        spec_clean = predict_spectrum(h, s, c, params)
        spectra[i, :] = spec_clean .+ noise_σ .* randn(rng, length(wavelengths))
    end

    wavelengths, spectra, compositions
end

function _sample_simplex(rng::AbstractRNG)
    # Sample uniformly from the 2-simplex
    x1, x2 = sort([rand(rng), rand(rng)])
    h = x1; s = x2 - x1; c = 1.0 - x2
    h, s, c
end

end
